### PR TITLE
[c10d] support dynamic shapes for all_to_all_single_autograd

### PR DESCRIPTION
**Pull request by xmfan**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #157521

`all_to_all_single_autograd` is not an op, all the code executed until the `all_to_all_single` dispatch is visible to the compiler. This means the `all_to_all_single_autograd` wrapper code must support symints in order to be traceable with dynamic shapes.

FIXES https://github.com/pytorch/pytorch/issues/157479

cc @H-Huang @awgu @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @chauhang @amjames

### CODE REVIEW

File: torch/csrc/distributed/c10d/Functional.cpp [hunk 1]
```diff
@@ -197,10 +197,19 @@ at::Tensor reduce_scatter_tensor(
 
 at::Tensor all_to_all_single(
     const at::Tensor& input,
-    std::vector<int64_t> output_split_sizes,
-    std::vector<int64_t> input_split_sizes,
+    c10::SymIntArrayRef _output_split_sizes,
+    c10::SymIntArrayRef _input_split_sizes,
     // NOLINTNEXTLINE(performance-unnecessary-value-param)
     std::string group_name) {
+  std::vector<int64_t> output_split_sizes;
```
Comment (Skylion007) [inline torch/csrc/distributed/c10d/Functional.cpp:204]: Reserve sizes of output vectors here.

File: test/dynamo/test_fake_distributed.py [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+        torch._dynamo.mark_dynamic(x, 1)
+        wait_tensor(fn(x))
+        self.assertEqual(len(backend.fw_graphs), 1)
+        self.assertEqual(len(backend.bw_graphs), 1)
+        self.assertExpectedInline(
+            normalize_graph(backend.fw_graphs[0]),
+            """\
+class GraphModule(torch.nn.Module):
+    def forward(self, primals_1: "Sym(s77)", primals_2: "Sym(s27)", primals_3: "f32[s77, s27]"):
+        floordiv: "Sym((s77//2))" = primals_1 // 2
```
Comment (wconstab) [inline test/dynamo/test_fake_distributed.py:57]: Just curious, does the normalize helper ensure these symint numberings are stable? Seems to me if someone adds another symint supported op that's used in this test case it would destabilize the numberings, though that's unlikely since the test case is just one op

File: test/dynamo/test_fake_distributed.py [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+        return (None, None, wait_tensor_1)
+""",  # noqa: B950
+        )
+
+        backend.fw_graphs.clear()
+        backend.bw_graphs.clear()
+
+        # Test unbacked shapes
+        x = torch.randn(8, 8, 8, requires_grad=True)
+        torch._dynamo.decorators.mark_unbacked(x, 0)
```
Comment (wconstab) [inline test/dynamo/test_fake_distributed.py:81]: For my education, mark unbacked is helpful for explicitness, but wouldn't unbacked normally occur implicitly e.g by tracing a graph with fake inputs that have no values? Why use real tensors at all here?

**Review by wconstab — APPROVED** [review]

File: test/dynamo/test_fake_distributed.py (hunk 2 above)
Comment (xmfan) [reply test/dynamo/test_fake_distributed.py:57]: it doesn't, if we did add another op into this test, i would want the author to regenerate the printed graph, and manually check that the symints are properly handled

File: test/dynamo/test_fake_distributed.py (hunk 3 above)
Comment (xmfan) [reply test/dynamo/test_fake_distributed.py:81]: The shapes of the fake tensor would still need to be static/dynamic/unbacked/etc. Fake tensor inputs only affect data-dependent operations.

In the wild, we generally graph break when we need an unbacked symint. An example is taking the `.item()` of some scalar tensor and branching on it. There's some flags to attempt their capture e.g. capture_scalar_outputs, mostly for export, which lead to using unbacked symints to represent the scalar value symbolically.

Another valid use of explicit mark unbacked outside of tests is to enforce a graph without any specialization on that shape, e.g. ensure no recompiles due to 0/1 shapes. This test is most representative for this last case.
